src/celengine/observer.cpp

```diff
@@ -976,42 +976,39 @@ bool Observer::orbit(const Selection& selection, const Eigen::Vector3f &from, co
  * at a rate dependent on the observer's distance from the object.
  */
 void Observer::changeOrbitDistance(const Selection& selection, float d)
+{
+    scaleOrbitDistance(selection, std::exp(-d));
+}
+
+
+void Observer::scaleOrbitDistance(const Selection& selection, float scale)
 {
     Selection center = frame->getRefObject();
-    if (center.empty() && !selection.empty())
+    if (center.empty())
     {
+        if (selection.empty())
+            return;
+
         center = selection;
         setFrame(frame->getCoordinateSystem(), center);
     }
 
-    if (!center.empty())
-    {
-        UniversalCoord focusPosition = center.getPosition(getTime());
-
-        double size = center.radius();
+    UniversalCoord focusPosition = center.getPosition(getTime());
 
-        // Somewhat arbitrary parameters to chosen to give the camera movement
-        // a nice feel.  They should probably be function parameters.
-        double minOrbitDistance = size;
-        double naturalOrbitDistance = 4.0 * size;
+    // Determine distance and direction to the selected object
+    Eigen::Vector3d v = getPosition().offsetFromKm(focusPosition);
+    double currentDistance = v.norm();
 
-        // Determine distance and direction to the selected object
-        Vector3d v = getPosition().offsetFromKm(focusPosition);
-        double currentDistance = v.norm();
+    double minOrbitDistance = center.radius();
+    if (currentDistance < minOrbitDistance)
+        minOrbitDistance = currentDistance * 0.5;
 
-        if (currentDistance < minOrbitDistance)
-            minOrbitDistance = currentDistance * 0.5;
+    double span = currentDistance - minOrbitDistance;
+    double newDistance = minOrbitDistance + span / static_cast<double>(scale);
+    v = v * (newDistance / currentDistance);
 
-        if (currentDistance >= minOrbitDistance && naturalOrbitDistance != 0)
```

Comment:
The only change to existing behavior is in the old code if radius == 0 (by checking naturalOrbitDistance != 0), distance would not be changed at all (which in my opinion does not make a lot of sense). In the new code, distance would still be scaled.